### PR TITLE
[WIP] Post-processing

### LINKED ISSUE #19: How to access the target cloned parent?
When doing a deep cloning of an ActiveRecord object, I would like a child to access it's cloned parent in the `finalize` block. Instead when I access the parent association, it is linked to the source parent.

Is this there a way for the child of deep cloned object to retrieve it's parent association which is also a cloned object?

### PR DESCRIPTION
**Pull request by ssnickolay**
- [x] Introduce `after_persist` feature (currently, supports only `ActiveRecord` adapter)
- [x] Introduce `Operation` as a result of cloning
- [x] Clean up `resolvers`, `declarations` and `adapters` layers
- [x] Rename test models to use a more meaningful domain model
- [x] Update Rubocop to `0.61`
- [x] Update bundler to `2.0`
- [x] Drop supporting for Ruby 2.2

-----

Dept:
- Pass calculated object over all post-processing blocks to control process (ex: validation case)
- Use `Operation` for `Sequel` adapter
- Update Docs
- Update Changelog

### CODE REVIEW

File: lib/clowne/utils/operation.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            clear!
+          end
+        end
+
+        def clear!
+          Thread.current[THREAD_KEY] = nil
+        end
+      end
+
+      attr_accessor :clone
```
Comment (palkan) [inline lib/clowne/utils/operation.rb]: `clone` could be confused with a `#clone` method, which present in many Ruby classes (especially, Rails); maybe, `clone_record` or `clone_result`?

File: lib/clowne/utils/operation.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+      def add_post_processing(block)
+        @post_processings.unshift(block)
+      end
+
+      def add_mapping(origin, clone)
+        @mapper.add(origin, clone)
+      end
+
+      def save
```
Comment (palkan) [inline lib/clowne/utils/operation.rb]: Maybe, we should introduce our own API?

`commit` / `commit!` sounds meaningful, or not?

We can add `save` / `save!` for backward compatibility and add a deprecation warnings.

File: lib/clowne/utils/operation.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        @clone.save
+      end
+
+      def save!
+        @clone.save!
+      end
+
+      def do_post_processing!
```
Comment (palkan) [inline lib/clowne/utils/operation.rb]: Is there a reason of calling `do_post_porcessing!` as a separate step and not the part of the `save!` call?

As I see in the [example](https://github.com/onboardiq/fountain/compare/chore/stage-cloner...chore/stage-cloner-2), we always call `do_post_processing!` after `save!`.

I suggest something like this:

```ruby
def commit!
  # wrap into transaction
  clone_result.transaction do
    clone_result.save!
    do_post_processing
  end
end

def commit
  clone_result.transaction do
    do_post_processing if clone_result.save
  end
end
```

File: lib/clowne/adapters/base.rb [hunk 4]
```diff
@@ -71,6 +37,23 @@ def init_record(record)
...
+  Clowne::Resolvers::Nullify
+)
+
+Clowne::Adapters::Base.register_resolver(
+  :finalize, Clowne::Resolvers::Finalize,
+  after: :nullify
+)
+
+Clowne::Adapters::Base.register_resolver(
+  :post_processing, Clowne::Resolvers::PostProcessing,
```
Comment (palkan) [inline lib/clowne/adapters/base.rb]: Maybe, rename it to `after_commit`? (if we go with `commit` naming)

File: lib/clowne/utils/operation.rb (hunk 1 above)
Comment (ssnickolay) [reply lib/clowne/utils/operation.rb]: Totally agree, I already had problems with this lousy naming ... `clone_record` the best for me

File: lib/clowne/utils/operation.rb (hunk 3 above)
Comment (ssnickolay) [reply lib/clowne/utils/operation.rb]: It can simply be implemented only for ActiveRecord :( Sequel (ROM) has another way to use transaction so it will depend on an adapter and I not sure that it is a great idea. And also I cannot to say that _ALL_ users want to execute `post_processing` with a transaction. Yep, its general approach but not the only one and to provide the more flexible solution (when user wrap to transaction outside gem) sounds better for me. For example, a user can save clone and accumulate errors from `post_processing` blocks. Or uses `fail fast` strategy and aborts all next `post_processing` blocks but keeps origin clone in DB. (all examples will be possible when I'll implement parameters for post-processing blocks).
And the last: `do_post_processing!` it is **really** dangerous operation which can change a lot of records and I think we should provide explicit execution for it.
Anyway, what do you think about "sugar" via kwargs:

```ruby
def commit!(do_post_processing: false)
# or def commit!(after_commit: false)
end
```
?

File: lib/clowne/adapters/base.rb (hunk 4 above)
Comment (ssnickolay) [reply lib/clowne/adapters/base.rb]: Hm...it's better than `post_processing `but how do you think will people not be confused with AR callback? 🤔

Comment (palkan) [reply lib/clowne/adapters/base.rb]: Yep, it could be.

Maybe, `after_clone`?

File: lib/clowne/cloner.rb [hunk 5]
```diff
@@ -93,8 +92,17 @@ def plan_with_traits(ids)
 
       private
 
+      def with_operation(options)
+        return yield unless adapter.is_a?(Clowne::Adapters::ActiveRecord) # TODO: for all adapters
```
[the remark this answers is not in the document]
Comment (ssnickolay) [reply lib/clowne/cloner.rb]: Yep, I wanted to do it the next step when I implement `Operation` for Sequel adapter...but I can to do it right now

File: lib/clowne/utils/operation.rb (hunk 3 above)
Comment (ssnickolay) [reply lib/clowne/utils/operation.rb]: > 👍 I like commit(skip_after_clone: true)

 Hm, your key point is execution `after_clone` is the default behavior, am I right?

File: lib/clowne/adapters/base.rb (hunk 4 above)
Comment (ssnickolay) [reply lib/clowne/adapters/base.rb]: `after_clone` very closest to `finalize` IMO :( Also not so readable: `after_clone do |origin, clone|`

Couple of Variants:

1. `fulfill_with_persisted { |origin, clone| .. }`
2. `with_persisted { |origin, clone| .. }`
3. `process_persisted { |origin, clone| .. }` 

Or something similar to `to fix` (because it is the main goal):
1. `repair { |origin, clone| ... }`
2. `overhaul { |origin, clone| ... }` 
3. `grind { |origin, clone| .. }`
4. `smooth { |origin, clone| .. }`

**Comment by ssnickolay** [comment]
```
clone record (the current `Cloner#call` behaviour) 
 |__ do something with the clone record (unsaved) (`after_clone` callback?)
 |__ persist record (`save`)
 |__ do something after saved (`after_persist` callback?)
```

So, we need two different callbacks: `after_clone` and `after_persist`. Make sense for me (both names and count of callback).

> In this case, I would suggest calling after_clone callbacks by default when calling Cloner#cloner (I don't even think we need an option to skip this, that should be a part of callback logic itself).

Good for me :+1:

> What about transactions? I see two options:
....

That's why I want to `move` all transaction handling to the user side. Just because in another way it could lead to several edge cases (especially with non AR adapters) which we cannot see for this moment. At least for the first implementation. But I understand your thoughts and transactions `out the box` can be useful...Maybe in future, we will add default transaction via kwargs or via other `Operation` interface but it not so clear for now, IMO. 

TL;DR our plan:
1. Rename `post_processing`  callback  to `after_persist`
2. Add `Operation#persist!(after_persist: true)` and `Operation#to_record` as we discussed above. Without transaction.
3. Add issue for `after_clone` callback (good for http://cultofmartians.com ;) )
4. Complete checkboxes in describing (README etc)

Wdyt?

**Comment by palkan** [comment]
LGTM.

Only one question:

> Add Operation#persist!(after_persist: true)

Do we need this `after_persist` modifier? What is the case?

**Comment by ssnickolay** [comment]
> Do we need this after_persist modifier? What is the case?
> ...
> And after_persist callback is called when we call CloneResult#persist.

I think we should provide a possibility to save the record without `after_persist` callbacks. 

I see two options for this:

1. Use kwargs and `persist!(after_safe: true) == persist!` # after_safe: true - by default
2. Provide 3 different methods:
- `save!` # just save without `after_persist` callbacks
- `run_after_persist!` # run only callbacks
- `persist!` # the same as `save! && run_after_persist!`

IMO the best decision to combinate both options:

```ruby
class Operation
  def persist!(after_safe: true)
   to_record.save!
   run_after_persist! if after_safe
  end

  def run_after_persist!
     raise unless to_record.persisted?
     # calls calbacks
  end
end
```

So user can use what he want:
```ruby
operation.persist! 
# the same as
operation.to_record.save! && operation.run_after_persist!

operation.persist!(after_persist: false)
# the same as
operation.to_record.save!
```

**Comment by palkan** [comment]
> I think we should provide a possibility to save the record without after_persist callbacks.

It sounds like calling `ActiveRecord#save!` without calling callbacks; if there is a callback, it means something, _one can not just skip it_.

Btw, do we pass `params` to callbacks? We could used them to control whether we want to run the callback or not:

```ruby
after_persist do |origin, record, skip_something: false|
  next if skip_something
  # ...
end
```

This way we move responsibility of calling callbacks to the cloner class itself.
